### ISSUE TITLE
Barbara should care about correctness

### ISSUE THREAD

**Issue by Diggsey** [rust-lang/wg-async-foundations#151]
At the moment, the benefits of "productivity" and "stability" are mentioned. I think an experienced Rust developer would care about "correctness" at least as much.

I think this is particularly relevent for async-foundations because other aspects of the language place a strong focus on correctness (eg. no iterator invalidation) and async/await is not great for this right now.

There are a lot of ways to shoot yourself in the foot with async/await:

- Calling an async function that takes a `&mut` parameter (even if it's `&mut self`) and interrupting it, will generally leave the parameter in an unrecoverable state, and yet there is no barrier to accessing it further.
- A widely used pattern is to share state between async code via an `Arc<Mutex<...>>` which can easily lead to race conditions, as the programmer expects the state to be accessed in one order, but due to yield points, the state may be accessed in a different order.
- Various rules like not holding a mutex across a yield point, or not blocking an executor, must be manually enforced.

To name a few.

I think there are already stories covering many of these things, I just think it's odd not to mention "correctness" as a motivating factor, as this is the primary reason I use Rust.

**Comment by nikomatsakis** [comment]
I think I might rather make that Grace's concern; it's kind of a generalization of "memory safety". The goal was that we cover the key attributes of Rust with one per character -- e.g., performance is Alan, correctness is Grace, etc -- but I think that the characters also being tied to language families may be causing some confusion. I may want to tweak that setup.

Regardless, I agree with you that correctness overall -- and not just memory safety -- is a key consideration for Rust programmers.

**Comment by uazu** [comment]
I agree. Provable correctness is what attracted me to Rust, and the more runtime checks there are going on, or other possible runtime failure modes (deadlocks etc), the less comfortable I become with the whole thing. I think strict checking is possible to some extent through the borrow checker and GhostCell-like techniques (e.g. maybe something like qcell's TLCell for multi-threaded). Effectively this means stuff at the top of the stack has the most privileges, but by tying up owner borrows and thus not being able to pass them on, those privileges can be statically reduced as we go down the stack. So code deeper down the stack can't do as much, and is forced to defer actions to code higher up the stack to handle. This avoids a number of bad code patterns within a single thread. This might also help with some of the cases of holding stuff over yield points, since a cell borrow has to be released in order to yield, as the yield must release the owner borrow as it returns back up the stack. However I don't know how you'd extend these compile-time cell checks into a multi-threaded context, e.g. behind `Arc<Mutex>` as you say. Maybe some kind of `Arc<Mutex<ZCell<T>>>` or `Arc<ZCell<Mutex<T>>>` could be devised, which guarantees that our code is the only code on this thread right now that is attempting access to the data within the mutex (where ZCell access depends on having a mutable borrow on a ZOwner held further up the stack, with a single owner on each thread). Would that be enough to avoid some of the bad patterns?
